src/portable/raspberrypi/rp2040/rp2040_usb.c

```diff
@@ -218,8 +246,14 @@ void hw_endpoint_xfer_start(struct hw_endpoint *ep, uint8_t *buffer, uint16_t to
   ep->active        = true;
   ep->user_buf      = buffer;
 
-  _hw_endpoint_start_next_buffer(ep);
-  _hw_endpoint_lock_update(ep, -1);
+  if(!rp2040_critical_frame_period(ep)) {
+    hw_endpoint_start_next_buffer(ep);
+  } else {
+#if TUD_OPT_RP2040_USB_DEVICE_UFRAME_FIX
+    pending_nonperiodic_xfer |= (uint16_t)(1 << tu_edpt_number(ep->ep_addr));
```

Comment:
`pending_nonperiodic_xfer` are both read-then-set in both here via dcd_edpt_xfer() and sof isr. There is a race condition e.g 
- we need to set with 0x02 mask, after the read here when pending_nonperiodic_xfer = 0x01
- sof isr occurred, process the 0x01 and clear the bit, sof return
- pending_nonperiodic_xfer is set with 0x03 therefore incorrectly schedule an redundant transfer.

Reply:
Accesses to this variable are now guarded by the endpoint lock (which implements a pico-sdk critical_section). Would you prefer folding the variable into the hw_endpoint array?

Reply:
thank you for adding the guard, I think we should move the variable to hw_endpoint array, since the tinyusb stack only submit 1 transfer per endpoint at a time. This will remove the need to run critical section which is an rather expensive function.